core/jvm/src/main/scala/fs2/interop/flow/StreamSubscription.scala

```diff
@@ -77,94 +96,96 @@ private[flow] final class StreamSubscription[F[_], A] private (
         .drain
 
     events
-      .race(canceled.get)
+      .race(canceled)
       .guaranteeCase {
         case Outcome.Succeeded(result) =>
           result.flatMap {
-            case Left(())  => onComplete // Events finished normally.
+            case Left(())  => F.delay(onComplete()) // Events finished normally.
             case Right(()) => F.unit // Events was canceled.
           }
-        case Outcome.Errored(ex) => onError(ex)
-        case Outcome.Canceled()  => cancelMe
+        case Outcome.Errored(ex) => F.delay(onError(ex))
+        case Outcome.Canceled()  => F.delay(onComplete())
```

Comment:
If the `run` is canceled what should we do?
Right now it is behaving as if the `Stream` finished normally... but I am not totally comfortable with that.

Reply:
`onError(new CancellationException)` ?

Reply:
Yeah we can do that, or we could call `cancel`.
My question is what are the right / best semantics?

So, if I have something like:

```scala
val data: Stream[IO, Foo] = ...
val thirdPartySubscriber: Subscriber[Foo] = ...
val killSwitch: F[Unit] = ...

data.subscribe(thirdPartySubscriber).race(killSwitch)
```

And then `killSwitch` is turned on.
The initial gut feeling is that, from the perspective of the `thirdPartySubscriber`, we just finished producing data; e.g. `onComplete`. This makes a lot of sense if `data` is just infinite and the `killSwitch` is the shutdown of the whole application. Nevertheless, if `CancellationException` is made public, this could be documented behavior.
Moreover, what if `killSwitch` is there to signal a failure in the application? We probably want to signal the downstream consumer that something wrong happened; e.g. `onError`
Finally, we may also assume that from the perspective of `subscribe` the `killSwitch` is just another way that _downstream_ has to signal cancellation; e.g. not different from `cancel`.

Thinking about it, I think I either prefer to use a public exception that downstream users may catch, or let users configure the behavior.

Reply:
I think in that case stream interruption should be used (for graceful closure) instead of cancellation.